src/Checkout/ExpressCheckout/Service/PayPalExpressCheckoutDataService.php

```diff
@@ -104,9 +115,26 @@ protected function getButtonLanguageSetting(): string
         return Settings::ECS_BUTTON_LANGUAGE_ISO;
     }
 
-    private function showPayLater(string $salesChannelId, AvailabilityContext $availabilityContext): bool
+    private function showPayLater(string $salesChannelId, AvailabilityContext $availabilityContext, ?ShopwareSalesChannelEvent $event): bool
     {
-        return $this->systemConfigService->getBool(Settings::ECS_SHOW_PAY_LATER, $salesChannelId)
+        return $this->canCheckoutFromPage($event)
+            && $this->systemConfigService->getBool(Settings::ECS_SHOW_PAY_LATER, $salesChannelId)
             && $this->payLaterMethodData->isAvailable($availabilityContext);
     }
+
+    private function getProductFromEvent(?ShopwareSalesChannelEvent $event): ?SalesChannelProductEntity
+    {
+        if ($event instanceof ProductPageLoadedEvent) {
+            return $event->getPage()->getProduct();
+        }
+
+        return null;
+    }
+
+    private function canCheckoutFromPage(?ShopwareSalesChannelEvent $event): bool
+    {
+        return $event instanceof ProductPageLoadedEvent
+            || $event instanceof CheckoutCartPageLoadedEvent
+            || $event instanceof OffcanvasCartPageLoadedEvent;
+    }
```

Comment:
Not sure about this. There are other pages like CheckoutRegisterPage, that should also work.
I think, the logic should be:
- `$addProductToCart === false` -> show based on availabilityContextFromCart
- `$addProductToCart === false && $event instanceof ProductPageLoadedEvent` -> show based on availabilityContextFromProduct
- `$addProductToCart === false && $event is something else` -> TBD, either not show any PayLater buttons or all

Reply:
Looks good now, but the `canCheckoutFromPage` can be removed